File: model-client/src/main/java/org/modelix/model/lazy/CLNode.java
package org.modelix.model.lazy;

import org.modelix.model.persistent.CPElementRef;
import org.modelix.model.persistent.CPNode;

import java.util.List;
import java.util.function.Function;
import java.util.stream.Stream;
import java.util.stream.StreamSupport;

public class CLNode extends CLElement {

  public CLNode(CLTree tree, CPNode data) {
    super(tree, data);
  }

  public CLNode(CLTree tree, long id, String concept, long parentId, String roleInParent, long[] childrenIds, String[] propertyRoles, String[] propertyValues, String[] referenceRoles, CPElementRef[] referenceTargets) {
    this(tree, CPNode.create(id, concept, parentId, roleInParent, childrenIds, propertyRoles, propertyValues, referenceRoles, referenceTargets));
  }

  @Override
  public CPNode getData() {
    return (CPNode) super.getData();
  }

  public IBulkQuery.Value<Iterable<CLNode>> getChildren(IBulkQuery bulkQuery) {
    return tree.resolveElements(getData().getChildrenIds(), bulkQuery).map(elements -> elements);
  }

  public IBulkQuery.Value<Iterable<CLNode>> getDescendants(final IBulkQuery bulkQuery, boolean includeSelf) {
    if (includeSelf) {
      return getDescendants(bulkQuery, false)
              .map(descendants -> Stream.concat(
                      Stream.of(this),
                      StreamSupport.stream(descendants.spliterator(), false)
              )::iterator);
    } else {
      return getChildren(bulkQuery).mapBulk(children -> {
        IBulkQuery.Value<Iterable<CLNode>> d = bulkQuery
                .map(children, child -> child.getDescendants(bulkQuery, true))
                .map(it -> it.stream().flatMap(n -> StreamSupport.stream(n.spliterator(), false))::iterator);
        return d;
      });
    }
  }

  public String getConcept() {
    return getData().getConcept();
  }
}

File: model-client/src/main/kotlin/org/modelix/model/lazy/CLNode.kt
package org.modelix.model.lazy

import org.modelix.model.persistent.CPElementRef
import org.modelix.model.persistent.CPNode
import java.util.function.Function
import java.util.stream.Stream
import java.util.stream.StreamSupport

class CLNode(tree: CLTree?, data: CPNode?) : CLElement(tree!!, data!!) {
    constructor(tree: CLTree?, id: Long, concept: String?, parentId: Long, roleInParent: String?, childrenIds: LongArray?, propertyRoles: Array<String?>?, propertyValues: Array<String?>?, referenceRoles: Array<String?>?, referenceTargets: Array<CPElementRef?>?) :
        this(
            tree,
            CPNode.create(
                id, concept, parentId, roleInParent, childrenIds!!,
                propertyRoles as Array<String?>, propertyValues as Array<String?>,
                referenceRoles as Array<String?>, referenceTargets as Array<CPElementRef?>
            )
        ) {}

    override fun getData(): CPNode? {
        return super.getData() as CPNode?
    }

    fun getChildren(bulkQuery: IBulkQuery?): IBulkQuery.Value<Iterable<CLNode>?>? {
        return (getTree() as CLTree).resolveElements(getData()!!.getChildrenIds().toList(), bulkQuery!!)!!
            .map(Function<List<CLNode?>?, Iterable<CLNode>?> { elements: List<CLNode?>? -> elements as Iterable<CLNode> })
    }

    fun getDescendants(bulkQuery: IBulkQuery, includeSelf: Boolean): IBulkQuery.Value<Iterable<CLNode>>? {
        return if (includeSelf) {
            getDescendants(bulkQuery, false)!!
                .map(
                    Function { descendants: Iterable<CLNode> ->
                        Iterable<CLNode> {
                            Stream.concat(
                                Stream.of(this),
                                StreamSupport.stream(descendants.spliterator(), false)
                            ).iterator()
                        }
                    }
                )
        } else {
            getChildren(bulkQuery)!!.mapBulk(
                Function { children: Iterable<CLNode>? ->
                    val d = bulkQuery
                        .map(children, Function { child: CLNode -> child.getDescendants(bulkQuery, true) })!!
                        .map(Function { it: List<Iterable<CLNode>>? -> Iterable<CLNode> { it!!.stream().flatMap { n: Iterable<CLNode> -> StreamSupport.stream(n.spliterator(), false) }.iterator() } })
                    d
                }
            )
        }
    }

    val concept: String?
        get() = getData()!!.concept
}